### ISSUE TITLE
Bug Report: vttablet hanging when running PlannedReparentShard

### ISSUE THREAD

**Issue by mattlord** [vitessio/vitess#9936]
### Overview of the Issue

During a PRS we sometimes see the tablet stuck waiting for something at this point in the logs:
```
I0317 04:56:50.637023    9434 rpc_replication.go:388] DemotePrimary
I0317 04:56:50.638689    9434 rpc_replication.go:438] DemotePrimary disabling query service
I0317 04:56:50.638703    9434 state_manager.go:214] Starting transition to PRIMARY Not Serving, timestamp: 2022-03-17 04:25:38.205679188 +0000 UTC
I0317 04:56:50.638773    9434 tablegc.go:212] TableGC: closing
...
```

Here's a sample debug blocking profile covering this: https://gist.github.com/derekperkins/dd6d54809a98b582c03909061e639766

### Reproduction Steps

We suspect that it involves:
1. Setting up many active vstreams: MoveTables,Reshard,OnlineDDL,Messaging
2. Doing PRS while those are active

### Binary Version

```sh
v13.0.0
```

### Operating System and Environment details

```sh
N/A
```

**Comment by rohit-nayak-ps** [comment]
@derekperkins, @GuptaManan100  is going to create a branch tomorrow with added logs, around tablet shutdown, to see if we can narrow down where the tablet hangs.

**Comment by mattlord** [comment]
@derekperkins after reading [your comments, attachments](https://github.com/vitessio/vitess/issues/8909), and thinking about it a bit more... I have a suspicion similar to yours in that I suspect PRS blocking is a symptom of another problem. I (also) suspect that for some reason the message table related queries are getting backlogged in the tablet's MySQL instance (perhaps the table size, structure, or queries used against it have changed over the recent months and vitess versions) — which would explain why you see no new messages generated even before the PRS (maybe that's WHY you do the PRS?) — which in turn causes the PRS to block during primary demotion waiting for those open statements/transactions to finish on the current/old primary during the demotion step. PRS will, by default, wait indefinitely for any running statements/transactions to finish on the current primary during demotion.

Note that with vreplication (branching, online DDL, etc) you can have an open transaction on the source tablet/mysqld that lasts minutes or even hours to maintain the consistent snapshot for a copy rows phase/cycle. 

This behavior was described and investigated in the past here:
 - https://github.com/vitessio/vitess/pull/6645
 - https://github.com/vitessio/vitess/issues/6982
 - https://vitess.io/docs/13.0/design-docs/vttablet/fast-transitions/
 
With PRS — at least during that time — on any failover you could have a partial outage of many seconds or many minutes as it would wait for running statements/transactions to complete and it was not uncommon to have 1 or more long running ones in an instance at any given time and if they did not finish within the PRS timeout then you could also be left w/o a healthy shard too (although I think we fixed _at least_ most of the issues related to not fully reverting the operation to the previous steady/healthy state).
 
Have you explored setting the [`-shutdown_grace_period`](https://vitess.io/docs/13.0/reference/programs/vttablet/) to a [non-zero value](https://github.com/vitessio/vitess/blob/ec8d09b1ba00b435c27a2deebef75fcfcdf5fb78/go/vt/vttablet/tabletserver/tabletenv/config.go#L216), e.g. 10 seconds or something? This way the tablet will forcefully close all of those open connections, after the grace period has ended, to allow the PRS to continue.

If you see the issue happen again, is it possible to grab the mysqld logs and `mysql> show full process list;` output from that mysqld instance when you see things hanging? 

Please let me know if there's anything I missed or misunderstood, or any new info comes to light. I apologize for the problems it's been causing!

**Comment by GuptaManan100** [comment]
I have added a lot of logging in https://github.com/vitessio/vitess/pull/9942. Could you guys take a look at the PR? Once it is okay, @derekperkins would you mind cherrypicking it.

I personally think that the issue @derekperkins is seeing is coming from a lot of blocking calls to `Subscribe` in messager Engine. For example - 
```
goroutine 1121445943 [semacquire, 1205 minutes]:
sync.runtime_SemacquireMutex(0x12074d7, 0x0, 0x1ebc580)
	runtime/sema.go:71 +0x25
sync.(*Mutex).lockSlow(0xc0006488c0)
	sync/mutex.go:138 +0x165
sync.(*Mutex).Lock(...)
	sync/mutex.go:81
vitess.io/vitess/go/vt/vttablet/tabletserver/messager.(*Engine).Subscribe(0xc001d78100, {0x26be3f0, 0xc001bc24b0}, {0xc001c2c8a0, 0x23}, 0x7f92e0766e60)
	vitess.io/vitess/go/vt/vttablet/tabletserver/messager/engine.go:122 +0x94
```

There are 207 results in the dump when searching for `messager.(*Engine).Subscribe`, which makes me think that there are a lot of blocked calls to Subscribe. Both the Subscribe and Close function lock the same mutex `mu` in the manager engine. Look at [Close](https://github.com/vitessio/vitess/blob/2b5f72c5dd0411c3ff9da4517c4a8d32c4f57fc5/go/vt/vttablet/tabletserver/messager/engine.go#L90) and [Subscribe](https://github.com/vitessio/vitess/blob/2b5f72c5dd0411c3ff9da4517c4a8d32c4f57fc5/go/vt/vttablet/tabletserver/messager/engine.go#L122). 

If we have a lot of outstanding Subscribe calls already blocked on the lock, then calling Close will just be one more process blocked on the lock. It will race with all the others to acquire it. If we assume that if there are even 50 Subscribe blocked calls. Then the chance that the Close function will acquire the Lock (assuming ties in races broken arbitrarily) will be 2%. 
Here I also think that the Subscribe function is holding the lock for a long time, since it calls Subscribe on the manager as well which intern has a lock. So we can be in a state where we are holding the outer lock, but waiting on the internal lock, which will also make the outer lock blocked for that long. 
[Here](https://github.com/vitessio/vitess/blob/2b5f72c5dd0411c3ff9da4517c4a8d32c4f57fc5/go/vt/vttablet/tabletserver/messager/engine.go#L123) we are unlocking the messager Engine mutex as a deferred call, while we call Subscribe on the manager [Here](https://github.com/vitessio/vitess/blob/2b5f72c5dd0411c3ff9da4517c4a8d32c4f57fc5/go/vt/vttablet/tabletserver/messager/engine.go#L131)
Within this call, we are acquiring another lock, local for the manager [here](https://github.com/vitessio/vitess/blob/2b5f72c5dd0411c3ff9da4517c4a8d32c4f57fc5/go/vt/vttablet/tabletserver/messager/message_manager.go#L378). 

I haven't looked at this code in too much detail to know why we have a mutex for the entire messager Engine and individual mutexes for the messager managers.

In any case, we are only acquiring the messager manager mutex while holding the messager Engine mutex, then isn't the messager manager mutex wasteful? Am I missing something?

**Comment by derekperkins** [comment]
Thanks so much for digging into this. A couple of things I'll clarify:
1. My testing was more recent than v13, it was based on the [vt_streamed_cols](https://github.com/vitessio/vitess/tree/vt_streamed_cols) branch. I suspected that we were encountering this bug, so I wanted to see if that fixed it https://github.com/vitessio/vitess/pull/9833
2. We rarely/never use MoveTables,Reshard,OnlineDDL, but we do have tens of message tables running, and several Materialize vstreams running. None of these deadlocks have happened while a copy phase was running

> which would explain why you see no new messages generated even before the PRS (maybe that's WHY you do the PRS?) 

That's exactly correct - messages stop processing for certain tables entirely and never recover, so we run PRS to restart processing.

> Have you explored setting the [-shutdown_grace_period](https://vitess.io/docs/13.0/reference/programs/vttablet/) to a [non-zero value](https://github.com/vitessio/vitess/blob/ec8d09b1ba00b435c27a2deebef75fcfcdf5fb78/go/vt/vttablet/tabletserver/tabletenv/config.go#L97), e.g. 10 seconds or something

TIL that was an option, so no. :) I'm willing to try that out. Definitely seems like a good idea, even if it does end up just masking the underlying problem

**Comment by derekperkins** [comment]
> There are 207 results in the dump when searching for messager.(*Engine).Subscribe, which makes me think that there are a lot of blocked calls to Subscribe

One thing I forgot to mention is that we only saw this happening in a single keyspace when using v13. That keyspace was sharded, while we run messages out of multiple other unsharded keyspaces that seemed fine, and they process messages at similar or higher rates. The keyspace that locks up more however does do a lot more autoscaling of consumers, so it is more likely to see contention on Subscribe/Unsubscribe. In the past, we implemented a feature that would shut down a consumer if it showed that there were messages to process, but none were being delivered, and that exacerbated the problem.

> Once it is okay, @derekperkins would you mind cherrypicking it.

I can cherry pick it and roll it out today. We ended up rolling back to v12 for the failing keyspace.

> I haven't looked at this code in too much detail to know why we have a mutex for the entire messager Engine and individual mutexes for the messager managers.

I'm not 100% sure either. @sougou wrote the original implementation, and @vmg attempted to fix #8909 by refactoring some of those locks in #9134. 

> I (also) suspect that for some reason the message table related queries are getting backlogged in the tablet's MySQL instance (perhaps the table size, structure, or queries used against it have changed over the recent months and vitess versions)

The structure and queries have been pretty consistent since we first started using messages. As I described in https://github.com/vitessio/vitess/issues/8909#issuecomment-1062573356, a recent increase in table size definitely contributed, and I found that MySQL wasn't using the index I thought it was, but applying a different index seemed to solve that problem. I'm still not super clear on exactly how vstream works, especially as messaging switches between live processing with it vs polling for new messages, so it's still possible that's an issue.

**Comment by derekperkins** [comment]
I'm rebuilding now on my [same branch](https://github.com/vitessio/vitess/tree/vt_streamed_cols), with #9942 cherry picked on top. Will deploy as soon as the build completes
https://hub.docker.com/repository/registry-1.docker.io/vitess/base/builds/c1a58d39-4848-4237-bc1f-f16512ec7948

**Comment by mattlord** [comment]
> update: it's deployed now. Most of our heavy usage starts at midnight UTC, so hopefully I'll have some logs by tomorrow

@derekperkins if it's possible to also grab the output of `mysql> show full processlist;` in the primary tablet's mysqld instance when you see the messages stop flowing from the shard that would be very helpful.

Thanks!

**Comment by derekperkins** [comment]
I've been actively trying to get it to break for 10 days now and finally got it. I feel very confident now that the trigger is number of unacked messages (time_next is not null), and thus likely to do with the polling mechanism. There are zero active queries in the processlist. The full dump is at the gist, but here is the key part of the logs after calling PlannedReparentShard.

```
rpc_replication.go:388] DemotePrimary
rpc_replication.go:438] DemotePrimary disabling query service
state_manager.go:214] Starting transition to PRIMARY Not Serving, timestamp: 2022-03-28 23:58:14.437817374 +0000 UTC
state_manager.go:496] Started execution of unserveCommon
state_manager.go:498] Finished execution of handleShutdownGracePeriod
state_manager.go:501] Started online ddl executor close
executor.go:299] onlineDDL Executor - Acquiring lock - initMutex
executor.go:301] onlineDDL Executor - Acquired lock - initMutex
executor.go:307] onlineDDL Executor - Stopping timer ticks
executor.go:309] onlineDDL Executor - Closing the conpool
pool.go:144] connpool - started execution of Close
pool.go:146] connpool - found the pool
pool.go:153] connpool - calling close on the pool
pool.go:155] connpool - acquiring lock
pool.go:157] connpool - acquired lock
pool.go:160] connpool - closing dbaPool
pool.go:162] connpool - finished execution of Close
executor.go:312] onlineDDL Executor - finished Close execution
state_manager.go:503] Finished online ddl executor close
state_manager.go:504] Started table garbage collector close
tablegc.go:205] TableGC - started execution of Close. Acquiring initMutex lock
tablegc.go:207] TableGC - acquired lock
tablegc.go:215] TableGC: closing
tablegc.go:219] TableGC - closing pool
pool.go:144] connpool - started execution of Close
pool.go:146] connpool - found the pool
pool.go:153] connpool - calling close on the pool
pool.go:155] connpool - acquiring lock
pool.go:157] connpool - acquired lock
pool.go:160] connpool - closing dbaPool
pool.go:162] connpool - finished execution of Close
tablegc.go:222] TableGC - finished execution of Close
state_manager.go:506] Finished table garbage collector close
state_manager.go:507] Started lag throttler close
throttler.go:266] Throttler - started execution of Close. Acquiring initMutex lock
throttler.go:268] Throttler - acquired initMutex lock
throttler.go:278] Throttler - finished suspending tickers
throttler.go:281] Throttler - closing pool
pool.go:144] connpool - started execution of Close
pool.go:146] connpool - found the pool
pool.go:153] connpool - calling close on the pool
pool.go:155] connpool - acquiring lock
pool.go:157] connpool - acquired lock
pool.go:160] connpool - closing dbaPool
pool.go:162] connpool - finished execution of Close
throttler.go:284] Throttler - finished execution of Close
state_manager.go:509] Finished lag throttler close
state_manager.go:510] Started messager close
engine.go:90] messager Engine - started execution of Close. Acquiring mu lock
```

https://gist.github.com/derekperkins/a2a9bc0d83a0589e21c8820c93e949e8

**Comment by derekperkins** [comment]
@sougou mentioned on the call that etcd might be involved. Here are the logs from all 5 pods during all the deadlocks and cutovers. The etcd pods restarted, but we ended up rolling new nodes multiple times, so I'm not sure if that was because they died or if they just restarted with nodes.
https://gist.github.com/derekperkins/9ec022994deb57d286f0ffcc80f516a4

**Comment by GuptaManan100** [comment]
So our initial hunch of the DemotePrimary thread being blocked on acquiring the loch for Messager close https://github.com/vitessio/vitess/issues/9936#issuecomment-1075109128 was correct. However, I don't see which process is currently holding that lock...

**Comment by derekperkins** [comment]
There also seems to be some sort of correlation with sharding, as we don't see this happening in our unsharded keyspaces, even at significantly higher load and row counts. Here's the rough outline of our data pipeline:

1. Unsharded keyspace loads 6M messages into a single table (those are generated and fan-out/propagate through several queues)
2. A queue worker then began adding those 6M messages to the next stage, which is a message table in a sharded keyspace (16 shards - ~400k messages / table). 
3. Normally the first sharded worker can easily keep up with the inbound flow (blue line newly created messages, green is acked messages) <img width="992" alt="image" src="https://user-images.githubusercontent.com/3588778/161451905-78b38080-8c2b-43dc-a71e-efce132478e8.png"> But when the message manager is deadlocked, queries and writes work, but some shards just don't deliver messages <img width="990" alt="image" src="https://user-images.githubusercontent.com/3588778/161451820-27cd9921-b39b-4cf0-81b6-2c05b38e5187.png">
4. There are several message tables in the sharded keyspace before fanning back into the unsharded keyspace, but even though that has 16x the volume, it also never deadlocks.

When a deadlock happens and messages stop happening, this is how I fix it:
1. Run `PlannedReparentShard` to the replica, which then fails due to the deadlock
2. Delete the k8s pod for the primary, which restarts in place
3. Run `PlannedReparentShard` again, which succeeds every time, but about 1/3 of the time, the original primary still thinks that it is the primary, so I delete the pod again, and then it restarts as a replica
4. At this point, messages reliably start running from the new primary

What is very interesting is that there seems to be shard interaction, which I wouldn't expect. If I run the steps above for multiple shards simultaneously, often none of them start working again, though it is 100% reliable to run one at a time. Sometimes I could get 2-3 shards fixed at the same time, but any more than that seemed to never work. I don't have any explanation for why that would be happening.

There is also some sort of threshold effect. It worked great for 10 days, then exploded over the weekend, with hundreds of deadlocks happening, and then once the message spike passed like a kidney stone, things are back to normal.

**Comment by derekperkins** [comment]
This happened again over the weekend and took 3-4 days of constant babysitting until it settled down. I am still suspicious of query timeouts / index issues. I started to do some `EXPLAIN` queries, and I think there might be an issue there. I ran both explains below on tables with a handful of rows and larger tables with millions of rows. It's not optimal that we are filtering on `WHERE < time_next` (pull in any messages that are ready to be processed), but `ORDER BY priority, time_next DESC` (handle high priority messages first, then start with more recent messages). This appears to render the ` (priority, time_next DESC)` index useless. I'm not sure how to reconcile that.

If we assume that a query timeout is responsible, @sougou left a warning that if the callback / send function doesn't return, Close will deadlock, which is exactly what we're seeing here. I followed that callback code upstream, but quickly got out of my depth.

https://github.com/vitessio/vitess/blob/8a88844e29b74adbba3c7f82525d2b5096aaab66/go/vt/vttablet/tabletserver/messager/engine.go#L123-L137

### Base query in messages ([link](https://github.com/vitessio/vitess/blob/main/go/vt/vttablet/tabletserver/messager/message_manager.go#L254-L256))

```sql
select priority, time_next, epoch, time_acked, id, keyspace_id 
from searches__extractor__msgs
where time_next < UNIX_TIMESTAMP(NOW())*1e9 
order by priority, time_next desc

id            | 1
select_type   | SIMPLE
table         | searches__extractor__msgs
partitions    | <null>
type          | range
possible_keys | time_next
key           | time_next
key_len       | 9
ref           | <null>
rows          | 1
filtered      | 100.0
Extra         | Using index condition; Using filesort
```

### Force Index  (priority, time_next DESC)

```sql
select priority, time_next, epoch, time_acked, id, keyspace_id 
from searches__extractor__msgs FORCE INDEX (priority__time_next) 
where time_next < UNIX_TIMESTAMP(NOW())*1e9 
order by priority, time_next desc

id            | 1
select_type   | SIMPLE
table         | searches__extractor__msgs
partitions    | <null>
type          | index
possible_keys | <null>
key           | priority__time_next
key_len       | 10
ref           | <null>
rows          | 545754
filtered      | 33.33
Extra         | Using where
```

### Table definition

```sql
CREATE TABLE `searches__extractor__msgs` (
  `id` bigint NOT NULL COMMENT 'use an event_id if you have one',
  `keyspace_id` bigint NOT NULL COMMENT 'usually workspace_id/account_id|user_id|keyword_id|domain_id|etc',
  `priority` tinyint NOT NULL DEFAULT '50' COMMENT 'lower number priorities process first',
  `epoch` smallint NOT NULL DEFAULT '0' COMMENT 'Vitess increments this each time it sends a message. Can be larger than try_count if a worker is restarted or otherwise does not get a chance to process a message',
  `try_count` tinyint unsigned NOT NULL DEFAULT '0' COMMENT 'how many times the message has been executed by a worker',
  `time_scheduled` bigint NOT NULL DEFAULT ((unix_timestamp(now()) * 1e9)) COMMENT 'epoch nanoseconds',
  `time_next` bigint DEFAULT (`time_scheduled`) COMMENT 'epoch nanoseconds',
  `time_acked` bigint DEFAULT NULL COMMENT 'epoch nanoseconds',
  `time_created` bigint NOT NULL DEFAULT ((unix_timestamp(now()) * 1e9)) COMMENT 'epoch nanoseconds',
  `is_debug` tinyint(1) NOT NULL DEFAULT '0' COMMENT 'turns on debug during message processing',
  `latest_tried_at` timestamp NULL DEFAULT NULL COMMENT 'latest time a message was executed',
  `latest_ack_type` tinyint DEFAULT NULL COMMENT 'latest ack type (ack=1|nack=2|fail=3|backoff=4|no-op=5|invalid=6)',
  `latest_error` varbinary(255) DEFAULT NULL COMMENT 'latest error encountered - is not set to NULL on ack',
  `latest_error_code` tinyint DEFAULT NULL COMMENT 'latest error ncode',
  `first_tried_at` timestamp NULL DEFAULT NULL COMMENT 'first time a message was executed',
  `first_ack_type` tinyint DEFAULT NULL COMMENT 'first ack type (ack=1|nack=2|fail=3|backoff=4|no-op=5|invalid=6)',
  `first_error` varbinary(255) DEFAULT NULL COMMENT 'first error encountered - is not set to NULL on ack',
  `first_error_code` tinyint DEFAULT NULL COMMENT 'first error ncode',
  `worker_func_duration_ms` int DEFAULT NULL COMMENT 'execution time in ms inside the worker function',
  `tx_func_duration_ms` int DEFAULT NULL COMMENT 'execution time in ms inside the tx function',
  `buffer_wait_duration_ms` int DEFAULT NULL COMMENT 'time in ms spent waiting in the worker buffer',
  `parent_queue` varbinary(63) DEFAULT NULL COMMENT 'queue a message propagated from if available',
  `parent_id` bigint DEFAULT NULL COMMENT 'id in the parent queue table - only set if parent_queue is',
  `trace_id` varbinary(63) DEFAULT NULL COMMENT 'trace id - correlates with traces and errors',
  PRIMARY KEY (`id`),
  KEY `ack_idx` (`time_acked`),
  KEY `priority__time_next` (`priority`,`time_next` DESC),
  KEY `time_next` (`time_next`),
  CONSTRAINT `fk__s_p_serps_pipeline_extractor_msgs__pipeline_id` FOREIGN KEY (`id`) REFERENCES `searches_providers_serps_pipeline` (`pipeline_id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb3 ROW_FORMAT=COMPRESSED COMMENT='vitess_message,vt_ack_wait=300,vt_min_backoff=300,vt_max_backoff=1800,vt_purge_after=604800,vt_batch_size=10,vt_cache_size=10000,vt_poller_interval=30,vt_stream_cols=id;keyspace_id'
```

_* note `vt_stream_cols=id;keyspace_id` as this is based on #9670. All it does is limit the number of columns in the base query, which is reflected above_

**Comment by derekperkins** [comment]
Here are a few explain plans for different limits. The limit didn't change the query plan for the non-force index queries.

### FORCE INDEX LIMIT 1,000
```json
{
  "query_block": {
    "select_id": 1,
    "cost_info": {
      "query_cost": "200811.35"
    },
    "ordering_operation": {
      "using_filesort": false,
      "table": {
        "table_name": "searches__extractor__msgs",
        "access_type": "index",
        "key": "priority__time_next",
        "used_key_parts": [
          "priority",
          "time_next"
        ],
        "key_length": "10",
        "rows_examined_per_scan": 1000,
        "rows_produced_per_join": 191229,
        "filtered": "33.33",
        "cost_info": {
          "read_cost": "143436.75",
          "eval_cost": "19122.95",
          "prefix_cost": "200811.35",
          "data_read_per_join": "132M"
        },
        "used_columns": [
          "id",
          "keyspace_id",
          "priority",
          "epoch",
          "time_next",
          "time_acked"
        ],
        "attached_condition": "(`keywords`.`searches__extractor__msgs`.`time_next` < <cache>((unix_timestamp(now()) * 1e9)))"
      }
    }
  }
}
```

### FORCE INDEX LIMIT 10,000
```json
{
  "query_block": {
    "select_id": 1,
    "cost_info": {
      "query_cost": "190892.35"
    },
    "ordering_operation": {
      "using_filesort": false,
      "table": {
        "table_name": "searches__extractor__msgs",
        "access_type": "index",
        "key": "priority__time_next",
        "used_key_parts": [
          "priority",
          "time_next"
        ],
        "key_length": "10",
        "rows_examined_per_scan": 10000,
        "rows_produced_per_join": 181783,
        "filtered": "33.33",
        "cost_info": {
          "read_cost": "136351.75",
          "eval_cost": "18178.38",
          "prefix_cost": "190892.35",
          "data_read_per_join": "126M"
        },
        "used_columns": [
          "id",
          "keyspace_id",
          "priority",
          "epoch",
          "time_next",
          "time_acked"
        ],
        "attached_condition": "(`keywords`.`searches__extractor__msgs`.`time_next` < <cache>((unix_timestamp(now()) * 1e9)))"
      }
    }
  }
}
```

**Comment by rohit-nayak-ps** [comment]
We happened to have a review session of the messager code earlier today by @sougou. He mentioned that the choice for DESC  in `ORDER BY priority, time_next DESC` was a opinionated decision that gave higher priority for the latest messages in the query resultset. He suggested that we think of changing it to the normal sort order (ASC) to avoid table scans in MySQL 5.7 (until 8.x becomes more universal). 

Given your additional observation about the `priority, time_next DESC` index not being used because of `time_next < %a` filter, what do you think of changing it to `ORDER BY priority, time_next` . (We would additionally need to modify the `Less()` of the cache module ...)

wdyt, does the reasoning make sense or are we missing something?  I guess the key question is if this change will affect your application semantics?

**Comment by derekperkins** [comment]
I don't particularly care about `ORDER BY priority, time_next` vs `ORDER BY priority, time_next DESC`. IIRC, @sougou's original reasoning was to keep better latency for newer messages for SLO purposes, which makes sense to me. It would also be pretty trivial to add as a table option. I'd probably lean towards leaving it as is or adding an option, especially since 5.7 is EOL next year, but again, no strong feelings.

Either way, I don't think it makes a difference for this issue, since we are running 8.0.27.

**Comment by mattlord** [comment]
I believe that I've figured it out:  https://gist.github.com/mattlord/2ebab2be24ba66918e81ca617f8da56c

### TL;DR:
So in [`unsubscribe()` we have the main mutex](https://github.com/vitessio/vitess/blob/c9853797bda8752202583695133c41a25d390930/go/vt/vttablet/tabletserver/messager/message_manager.go#L421-L423), and while holding it we then [call `stopVStream()` if the subscriber count is at 0](https://github.com/vitessio/vitess/blob/c9853797bda8752202583695133c41a25d390930/go/vt/vttablet/tabletserver/messager/message_manager.go#L437-L439). Then [`stopVStream()` tries to get the stream mutex](https://github.com/vitessio/vitess/blob/c9853797bda8752202583695133c41a25d390930/go/vt/vttablet/tabletserver/messager/message_manager.go#L629-L631).

Currently, `runPoller()` — which by default, or more accurately when using [the documented message table structure here](https://vitess.io/docs/13.0/reference/features/messaging/#creating-a-message-table) via the `vt_poller_interval` comment directive,  the poller runs on a 30 second ticker — [gets the stream mutex first](https://github.com/vitessio/vitess/blob/c9853797bda8752202583695133c41a25d390930/go/vt/vttablet/tabletserver/messager/message_manager.go#L751-L758), then it [later tries to get the main mutex](https://github.com/vitessio/vitess/blob/c9853797bda8752202583695133c41a25d390930/go/vt/vttablet/tabletserver/messager/message_manager.go#L776-L778).

Boom, we are deadlocked. They're both waiting on each other.

This then in turn causes the PRS to hang because [during the transition from primary serving we stop the message service](https://github.com/vitessio/vitess/blob/c9853797bda8752202583695133c41a25d390930/go/vt/vttablet/tabletserver/state_manager.go#L450), which in turn [will wait trying to get the main mutex](https://github.com/vitessio/vitess/blob/c9853797bda8752202583695133c41a25d390930/go/vt/vttablet/tabletserver/messager/message_manager.go#L346-L353).

**Comment by derekperkins** [comment]
Great detective work, thanks for digging in @mattlord!
